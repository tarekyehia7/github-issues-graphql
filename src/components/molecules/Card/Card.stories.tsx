import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { Card, CardType } from './Card';

export default {
	title: 'molecules/Card',
	component: Card,
} as ComponentMeta<typeof Card>;

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const Primary: ComponentStory<typeof Card> = args => {
	return <Card {...args}>{args.children}</Card>;
};

Primary.args = {
	title: <>This is card title</>,
	type: CardType.withAvatar,
	authorUrl: 'https://github.com/tarekyehia7',
	avatarUrl: 'https://avatars.githubusercontent.com/u/11144528?v=4',
	children: <h4 style={{ margin: '1rem' }}>This is card body</h4>,
};
