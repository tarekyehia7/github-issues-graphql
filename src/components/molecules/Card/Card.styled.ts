import styled from 'styled-components';

import { Image } from '../../atoms/Image/Image';
import { Title } from '../../atoms/Title/Title';

export const Box = styled.div<{ withAvatar: boolean }>`
	color: ${({ theme }) => theme.colors.lightBlack};
	background-color: ${({ theme }) => theme.colors.white};
	border: 1px solid ${({ theme }) => theme.colors.lightGray};
	border-radius: 6px;
	position: relative;
	${({ withAvatar }) => withAvatar && 'margin: 4rem;'}
	${({ withAvatar }) => !withAvatar && 'margin-top: 1rem;'}
	margin-bottom: ${({ theme }) => theme.margins.margin4};
	margin-right: 0;
	@media (max-width: 480px) {
		& > a {
			display: none;
		}
		margin: 0rem;
		margin-top: ${({ theme }) => theme.margins.margin4};
	}
	${({ withAvatar, theme }) =>
		withAvatar &&
		`
        @media (min-width: 480px) {
            &:before {
                position: absolute;
                top: 11px;
                right: 100%;
                left: -8px;
                display: block;
                width: 8px;
                height: 16px;
                pointer-events: none;
                content: ' ';
                clip-path: polygon(0 50%, 100% 0, 100% 100%);
                background-color: ${theme.colors.lightGray};
            }
        }
    `}
`;

export const Body = styled.div`
	font-size: 14px;
	color: ${({ theme }) => theme.colors.black};
	@media (max-width: 480px) {
		overflow: auto;
	}
`;

export const ImageStyled = styled(Image)`
	position: absolute;
	left: -53px;
`;

export const TitleStyled = styled.div`
    padding: ${({ theme }) => theme.margins.margin4};
    background-color: #f6f8fa;
    border-top-left-radius: 6px;
    border-top-right-radius: 6px;
`;