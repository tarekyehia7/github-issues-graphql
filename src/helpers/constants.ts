export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};