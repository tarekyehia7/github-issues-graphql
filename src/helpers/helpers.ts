import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
