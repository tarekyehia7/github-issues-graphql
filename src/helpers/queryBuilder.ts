import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};