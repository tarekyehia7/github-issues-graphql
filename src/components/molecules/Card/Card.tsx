import React, { PropsWithChildren } from 'react';

import { Link, LinkType } from '../../atoms/link/Link';
import { Title, TitleSizeTypes } from '../../atoms/Title/Title';
import { AvatarImg, Body, Box } from './Card.styled';

export enum CardType {
	Normal,
	withAvatar,
}

type CardProps = {
	title: JSX.Element;
	type: CardType;
	authorUrl?: string;
	avatarUrl?: string;
};

export const Card = ({
	title,
	children,
	type,
	avatarUrl,
	authorUrl,
}: CardProps & PropsWithChildren) => {
	return (
		<Box withAvatar={type === CardType.withAvatar}>
			{type === CardType.withAvatar && (
				<Link type={LinkType.Normal} to={authorUrl}>
					<AvatarImg alt="avatar url" src={avatarUrl} />
				</Link>
			)}
			<Title titleSize={TitleSizeTypes.Small}>{title}</Title>
			<Body>{children}</Body>
		</Box>
	);
};
