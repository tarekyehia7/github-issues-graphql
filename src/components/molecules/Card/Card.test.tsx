import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { Card, CardType, CardProps } from './Card';
import { PageWithTheme } from '../../../helpers/testing/helpers';

const title = 'This is card title';
const props: CardProps = {
	title: <>{title}</>,
	authorUrl: 'https://github.com/tarekyehia7',
	avatarUrl: 'https://avatars.githubusercontent.com/u/11144528?v=4',
	children: <h4>This is card body</h4>,
	type: CardType.Normal,
};

const renderPage = (cardType: CardType) => {
	const { container, getByText, getByTestId } = render(
		<PageWithTheme>
			<Card {...props} type={cardType}>
				{props.children}
			</Card>
		</PageWithTheme>,
	);

	return {
		container,
		getByText,
		getByTestId,
	};
};

describe('<Card />', () => {
	it('Should matches snapshot', async () => {
		const { container } = renderPage(CardType.Normal);
		expect(container).toMatchSnapshot();
	});

	it('should not have avatar', async () => {
		const { getByTestId } = renderPage(CardType.Normal);

		expect(getByTestId('card').firstChild).not.toHaveProperty('href');
	});

	it('should not have avatar', async () => {
		const { getByTestId } = renderPage(CardType.withAvatar);

		expect(getByTestId('card').firstChild).toHaveProperty('href');
		expect(getByTestId('card').firstChild).toHaveProperty('href', props.authorUrl);
	});

	it('should have correct title', async () => {
		const { getByText } = renderPage(CardType.withAvatar);

		expect(getByText(title)).toBeInTheDocument();
	});
});
