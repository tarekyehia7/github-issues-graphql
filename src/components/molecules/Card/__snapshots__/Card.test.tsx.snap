// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Should matches snapshot 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1rem;
}

.c0 {
  color: #24292f;
  background-color: #ffffff;
  border: 1px solid #d0d7de;
  border-radius: 6px;
  position: relative;
  margin-top: 1rem;
  margin-bottom: 1rem;
  margin-right: 0;
}

.c3 {
  font-size: 14px;
  overflow: auto;
  color: #000000;
}

.c1 {
  padding: 1rem;
  background-color: #f6f8fa;
  border-top-left-radius: 6px;
  border-top-right-radius: 6px;
}

@media (max-width:480px) {
  .c0 {
    margin: 0rem;
    margin-top: 1rem;
  }

  .c0 > a {
    display: none;
  }
}

@media (max-width:480px) {
  .c3 {
    overflow: auto;
  }
}

<div>
  <div
    class="c0"
    data-testid="card"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        This is card title
      </div>
    </div>
    <div
      class="c3"
    >
      <h4>
        This is card body
      </h4>
    </div>
  </div>
</div>
`;
